File: benchmarks/binary-trees/binary_trees.c
/**
 * The Computer Language Benchmarks Game
 * https://salsa.debian.org/benchmarksgame-team/benchmarksgame/
 *
 * binary-trees benchmark in C
 */

#include <stdlib.h>
#include <stdio.h>
#include <stdbool.h>

typedef struct tree_node {
    struct tree_node* left;
    struct tree_node* right;
    int item;
} tree_node;

tree_node* make_tree(int depth) {
    tree_node* node = malloc(sizeof(tree_node));
    
    if (depth > 0) {
        node->left = make_tree(depth - 1);
        node->right = make_tree(depth - 1);
        node->item = depth;
    } else {
        node->left = NULL;
        node->right = NULL;
        node->item = 0;
    }
    
    return node;
}

int check_tree(tree_node* node) {
    if (node->left == NULL)
        return node->item;
    else
        return node->item + check_tree(node->left) - check_tree(node->right);
}

void delete_tree(tree_node* node) {
    if (node->left != NULL) {
        delete_tree(node->left);
        delete_tree(node->right);
    }
    free(node);
}

int main(int argc, char* argv[]) {
    int n = argc > 1 ? atoi(argv[1]) : 10;
    int min_depth = 4;
    int max_depth = n < min_depth + 2 ? min_depth + 2 : n;
    
    // Create and check stretch tree
    int stretch_depth = max_depth + 1;
    tree_node* stretch_tree = make_tree(stretch_depth);
    printf("stretch tree of depth %d\t check: %d\n", 
           stretch_depth, check_tree(stretch_tree));
    delete_tree(stretch_tree);
    
    // Create long-lived tree
    tree_node* long_lived_tree = make_tree(max_depth);
    
    // Create, check, and delete multiple trees
    for (int depth = min_depth; depth <= max_depth; depth += 2) {
        int iterations = 1 << (max_depth - depth + min_depth);
        int check = 0;
        
        for (int i = 1; i <= iterations; i++) {
            tree_node* temp_tree = make_tree(depth);
            check += check_tree(temp_tree);
            delete_tree(temp_tree);
        }
        
        printf("%d\t trees of depth %d\t check: %d\n", 
               iterations, depth, check);
    }
    
    // Check and delete long-lived tree
    printf("long lived tree of depth %d\t check: %d\n", 
           max_depth, check_tree(long_lived_tree));
    delete_tree(long_lived_tree);
    
    return 0;
}

File: benchmarks/binary-trees/binary_trees.rs
// The Computer Language Benchmarks Game
// https://salsa.debian.org/benchmarksgame-team/benchmarksgame/
//
// binary-trees benchmark in Rust

use std::env;

struct Node {
    left: Option<Box<Node>>,
    right: Option<Box<Node>>,
    item: i32,
}

impl Node {
    fn new(item: i32) -> Self {
        Node {
            left: None,
            right: None,
            item,
        }
    }

    fn new_with_children(item: i32, left: Node, right: Node) -> Self {
        Node {
            left: Some(Box::new(left)),
            right: Some(Box::new(right)),
            item,
        }
    }

    fn check(&self) -> i32 {
        match (&self.left, &self.right) {
            (None, None) => self.item,
            (Some(left), Some(right)) => self.item + left.check() - right.check(),
            _ => unreachable!(),
        }
    }
}

fn make_tree(depth: i32) -> Node {
    if depth <= 0 {
        return Node::new(0);
    }
    Node::new_with_children(depth, make_tree(depth - 1), make_tree(depth - 1))
}

fn main() {
    let n = env::args()
        .nth(1)
        .and_then(|n| n.parse().ok())
        .unwrap_or(10);
    let min_depth = 4;
    let max_depth = if min_depth + 2 > n { min_depth + 2 } else { n };

    // Create and check stretch tree
    let stretch_depth = max_depth + 1;
    let stretch_tree = make_tree(stretch_depth);
    println!(
        "stretch tree of depth {}\t check: {}",
        stretch_depth,
        stretch_tree.check()
    );

    // Create long-lived tree
    let long_lived_tree = make_tree(max_depth);

    // Create, check, and delete multiple trees
    for depth in (min_depth..=max_depth).step_by(2) {
        let iterations = 1 << (max_depth - depth + min_depth);
        let mut check = 0;

        for _ in 1..=iterations {
            let temp_tree = make_tree(depth);
            check += temp_tree.check();
        }

        println!("{}\t trees of depth {}\t check: {}", iterations, depth, check);
    }

    // Check long-lived tree
    println!(
        "long lived tree of depth {}\t check: {}",
        max_depth,
        long_lived_tree.check()
    );
}
